util/regexp-assemble/tests/preprocessor_test.py

```diff
@@ -631,3 +631,55 @@ def test_includes_content(self, context, include_file_name, include_file):
         output = assembler._run(Peekerator(contents.splitlines()))
 
         assert output == '(?:file contents|next line)'
+
+    def test_includes_content_recursive(self, context):
+        name1 = f'{uuid.uuid4()}.data'
+        name2 = f'{uuid.uuid4()}.data'
+
+        file1_path = PurePath.joinpath(context.include_files_directory, name1)
+        file2_path = PurePath.joinpath(context.include_files_directory, name2)
+
+        contents = rf'''##!> include {Path(file1_path).stem}
+first text 1'''
+        Path(file1_path).write_text(rf'''##!> include {Path(file2_path).stem}
+second text 5''')
+        Path(file2_path).write_text(rf'''third text 10''')
+
+        assembler = Assembler(context)
+
+        output = assembler._run(Peekerator(contents.splitlines()))
+
+        # Unlink before fail
+        Path.unlink(file1_path)
+        Path.unlink(file2_path)
+
+        assert output == '(?:second text 5|third text 10|first text 1)'
+
+    def test_includes_content_with_templates_recursive(self, context):
+        name1 = f'{uuid.uuid4()}.data'
+        name2 = f'{uuid.uuid4()}.data'
+
+        file1_path = PurePath.joinpath(context.include_files_directory, name1)
+        file2_path = PurePath.joinpath(context.include_files_directory, name2)
+
+        contents = rf'''##!> include {Path(file1_path).stem}
+first text 1'''
+        Path(file1_path).write_text(rf'''##!> include {Path(file2_path).stem}
+second text 5
+##!> template slashes [/\]
+{{{{slashes}}}}
+''')
+        Path(file2_path).write_text(r'''##!> template hex [a-fA-F0-9]
+{{hex}}
+third text
+''')
+
+        assembler = Assembler(context)
+
+        output = assembler._run(Peekerator(contents.splitlines()))
+
+        # Unlink before fail
+        Path.unlink(file1_path)
+        Path.unlink(file2_path)
+
+        assert output == '(?:second text 5|first text 1|[a-fA-F0-9]|third text|[\\/\\])'
```

Comment:
These tests are great! Please add one more that test that a template defined in the outermost scope will be applied to an include. I don't think that will work at the moment, as the template will be processed before the include. I think we need to order them explicitly:
1. run all includes
2. run all templates
3. run the normal assembly

Reply:
Added test `test_template_with_usage_included_from_other_file` which cover that case. For now the templates need to be defined (and included) first. The flow probably is to include the templates first and then all the rest, but we might think on a case there we have usages first and templates afterwards?

👉 I think having a "common template include file" could be the use case here, and this is working.

Regarding the order, I moved to phase 1-2-3 for includes, templates and normal assembly. Maybe we want to rename them back to `process_includes`, `process_templates` and `process_assembly`, but let me know what you think.

Reply:
Yes, I agree. If you include a file with a template, then it will work for any lines after it.

I like the phases, but I agree, for now it's clearer to use the processor names again. If we ever have enough things to move to "phases", we can do that then.

Reply:
Excellent. Renamed.